optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/DroolsConstraintFactory.java

```diff
@@ -0,0 +1,121 @@
+
+    private final SolutionDescriptor<Solution_> solutionDescriptor;
+    private final String defaultConstraintPackage;
+
+    public DroolsConstraintFactory(SolutionDescriptor<Solution_> solutionDescriptor) {
+        this.solutionDescriptor = solutionDescriptor;
+        ConstraintConfigurationDescriptor<Solution_> configurationDescriptor
+                = solutionDescriptor.getConstraintConfigurationDescriptor();
+        if (configurationDescriptor == null) {
+            defaultConstraintPackage = solutionDescriptor.getSolutionClass().getPackage().getName();
+        } else {
+            defaultConstraintPackage = configurationDescriptor.getConstraintPackage();
+        }
+    }
+
+    @Override
+    public <A> UniConstraintStream<A> fromUnfiltered(Class<A> fromClass) {
+        return new DroolsFromUniConstraintStream<>(this, fromClass);
+    }
+
+    @Override
+    public <A> BiConstraintStream<A, A> fromUniquePair(Class<A> fromClass, BiJoiner<A, A> joiner) {
+        throw new UnsupportedOperationException();
+    }
+
+    // ************************************************************************
+    // SessionFactory creation
+    // ************************************************************************
+
+    @Override
+    public ConstraintSessionFactory<Solution_> buildSessionFactory(Constraint[] constraints) {
+        ModelImpl model = new ModelImpl();
+
+        AbstractScoreHolder<?> scoreHolder = (AbstractScoreHolder<?>) solutionDescriptor.getScoreDefinition()
+                .buildScoreHolder(false);
+        Class<? extends AbstractScoreHolder> scoreHolderClass = scoreHolder.getClass();
+        Global<? extends AbstractScoreHolder> scoreHolderGlobal = globalOf(scoreHolderClass,
+                solutionDescriptor.getSolutionClass().getPackage().getName(),
+                DroolsScoreDirector.GLOBAL_SCORE_HOLDER_KEY);
+        model.addGlobal(scoreHolderGlobal);
+
+        List<DroolsConstraint<Solution_>> droolsConstraintList = new ArrayList<>(constraints.length);
+        Set<String> constraintIdSet = new HashSet<>(constraints.length);
+        for (Constraint constraint : constraints) {
+            if (constraint.getConstraintFactory() != this) {
+                throw new IllegalStateException("The constraint (" + constraint.getConstraintId()
+                        + ") must be created from the same constraintFactory.");
+            }
+            boolean added = constraintIdSet.add(constraint.getConstraintId());
+            if (!added) {
+                throw new IllegalStateException(
+                        "There are 2 constraints with the same constraintName (" + constraint.getConstraintName()
+                                + ") in the same constraintPackage (" + constraint.getConstraintPackage() + ").");
+            }
+            DroolsConstraint<Solution_> droolsConstraint = (DroolsConstraint) constraint;
```

Comment:
soft suggestion: instead of casting to `DroolsConstraint`, we could make the `Constraint[]` a type parameter in the abstract class and get `DroolsConstraint[]` directly here.